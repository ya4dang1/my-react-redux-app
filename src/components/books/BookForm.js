import React from "react";
import PropTypes from "prop-types";
import TextInput from "../common/TextInput";
import SelectInput from "../common/SelectInput";

const BookForm = ({
  book,
  authors,
  onSave,
  onChange,
  saving = false,
  errors = {}
}) => {
  return (
    <form onSubmit={onSave}>
      <h2>{book.id ? "Edit Book" : "Add Book"}</h2>
      {errors.onSave && (
        <div className="alert alert-danger" role="alert">
          {errors.onSave}
        </div>
      )}
      <TextInput
        name="title"
        label="Title"
        value={book.title}
        onChange={onChange}
        error={errors.title}
      />

      <SelectInput
        name="authorId"
        label="Author"
        value={book.authorId || ""}
        defaultOption="Select Author"
        options={authors.map(author => ({
          value: author.id,
          text: author.name
        }))}
        onChange={onChange}
        error={errors.author}
      />

      <TextInput
        name="category"
        label="Category"
        value={book.category}
        onChange={onChange}
        error={errors.category}
      />

      <button type="submit" disabled={saving} className="btn btn-primary">
        {saving ? "Saving..." : "Save"}
      </button>
    </form>
  );
};

BookForm.propTypes = {
  authors: PropTypes.arrayOf(PropTypes.object).isRequired,
  book: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.number, PropTypes.string])
  ).isRequired,
  errors: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.object, PropTypes.string])
  ),
  onSave: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  saving: PropTypes.bool
};

BookForm.defaultProps = {
  errors: {},
  saving: null
};

export default BookForm;
