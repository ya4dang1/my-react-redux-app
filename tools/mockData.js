const books = [
  {
    id: 1,
    title: "Harry Potter and the Philosopher's Stone",
    slug: "harry-potter-ps",
    authorId: 1,
    category: "British Fantasy"
  },
  {
    id: 2,
    title: "Harry Potter and the Chamber of Secrets",
    slug: "harry-potter-cs",
    authorId: 1,
    category: "British Fantasy"
  },
  {
    id: 3,
    title: "Harry Potter and the Prisoner of Azkaban",
    slug: "harry-potter-pa",
    authorId: 1,
    category: "British Fantasy"
  },
  {
    id: 4,
    title: "Harry Potter and the Globet of Fire",
    slug: "harry-potter-gf",
    authorId: 1,
    category: "British Fantasy"
  },
  {
    id: 5,
    title: "Harry Potter and the Order of Phoenix",
    slug: "harry-potter-op",
    authorId: 1,
    category: "British Fantasy"
  },
  {
    id: 6,
    title: "Harry Potter and the Half-Blood Prince",
    slug: "harry-potter-hbp",
    authorId: 1,
    category: "British Fantasy"
  },
  {
    id: 7,
    title: "Harry Potter and the Deathly Hallows",
    slug: "harry-potter-dh",
    authorId: 1,
    category: "British Fantasy"
  },
  {
    id: 8,
    title: "A Game of Thrones",
    slug: "game-of-thrones-gt",
    authorId: 2,
    category: "American Fantasy"
  },
  {
    id: 9,
    title: "A Clash of Kings",
    slug: "game-of-thrones-ck",
    authorId: 1,
    category: "American Fantasy"
  },
  {
    id: 10,
    title: "A Storm of Swords",
    slug: "game-of-thrones-sw",
    authorId: 1,
    category: "American Fantasy"
  }
];

const authors = [
  { id: 1, name: "J. K. Rowling" },
  { id: 2, name: "George R. R. Martin" },
  { id: 3, name: "Adam Tang" }
];

const newBook = {
  id: null,
  title: "",
  authorId: null,
  category: ""
};

// Using CommonJS style export so we can consume via Node (without using Babel-node)
module.exports = {
  newBook,
  books,
  authors
};
