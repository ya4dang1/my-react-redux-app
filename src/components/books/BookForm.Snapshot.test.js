import React from "react";
import renderer from "react-test-renderer";
import BookForm from "./BookForm";
import { books, authors } from "../../../tools/mockData";

it("sets submit button label 'Saving...' when saving is true", () => {
  const tree = renderer.create(
    <BookForm
      book={books[0]}
      authors={authors}
      onSave={jest.fn()}
      onChange={jest.fn()}
      saving
    />
  );

  expect(tree).toMatchSnapshot();
});

it("sets submit button label 'Save' when saving is false", () => {
  const tree = renderer.create(
    <BookForm
      book={books[0]}
      authors={authors}
      onSave={jest.fn()}
      onChange={jest.fn()}
      saving={false}
    />
  );

  expect(tree).toMatchSnapshot();
});
