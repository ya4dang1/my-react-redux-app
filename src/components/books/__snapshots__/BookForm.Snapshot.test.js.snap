// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sets submit button label 'Save' when saving is false 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <h2>
    Edit Book
  </h2>
  <div
    className="form-group"
  >
    <label
      htmlFor="title"
    >
      Title
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="title"
        onChange={[MockFunction]}
        placeholder=""
        type="text"
        value="Harry Porter and the Philosopher's Stone"
      />
      
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="authorId"
    >
      Author
    </label>
    <div
      className="field"
    >
      <select
        className="form-control"
        name="authorId"
        onChange={[MockFunction]}
        value={1}
      >
        <option
          value=""
        >
          Select Author
        </option>
        <option
          value={1}
        >
          J. K. Rowling
        </option>
        <option
          value={2}
        >
          George R. R. Martin
        </option>
        <option
          value={3}
        >
          Adam Tang
        </option>
      </select>
      
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="category"
    >
      Category
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="category"
        onChange={[MockFunction]}
        placeholder=""
        type="text"
        value="British Fantasy"
      />
      
    </div>
  </div>
  <button
    className="btn btn-primary"
    disabled={false}
    type="submit"
  >
    Save
  </button>
</form>
`;

exports[`sets submit button label 'Saving...' when saving is true 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <h2>
    Edit Book
  </h2>
  <div
    className="form-group"
  >
    <label
      htmlFor="title"
    >
      Title
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="title"
        onChange={[MockFunction]}
        placeholder=""
        type="text"
        value="Harry Porter and the Philosopher's Stone"
      />
      
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="authorId"
    >
      Author
    </label>
    <div
      className="field"
    >
      <select
        className="form-control"
        name="authorId"
        onChange={[MockFunction]}
        value={1}
      >
        <option
          value=""
        >
          Select Author
        </option>
        <option
          value={1}
        >
          J. K. Rowling
        </option>
        <option
          value={2}
        >
          George R. R. Martin
        </option>
        <option
          value={3}
        >
          Adam Tang
        </option>
      </select>
      
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="category"
    >
      Category
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="category"
        onChange={[MockFunction]}
        placeholder=""
        type="text"
        value="British Fantasy"
      />
      
    </div>
  </div>
  <button
    className="btn btn-primary"
    disabled={true}
    type="submit"
  >
    Saving...
  </button>
</form>
`;
